hn1.1
.TH HN1 1 2009-03-23 htools "Ganeti H-tools"
.SH NAME
hn1 \- N+1 fixer for Ganeti

.SH SYNOPSIS
.B hn1
.B "[-C]"
.B "[-p]"
.B "[-o]"
.BI "[ -m " cluster "]"
.BI "[-n " nodes-file " ]"
.BI "[ -i " instances-file "]"
.BI "[-d " depth "]"
.BI "[-r " max-removals "]"
.BI "[-L " max-delta "]"
.BI "[-l " min-delta "]"

.B hn1
.B --version

.SH DESCRIPTION
hn1 is a cluster N+1 fixer that tries to compute the minimum number of
moves needed for getting all nodes to be N+1 compliant.

The algorithm is designed to be a 'perfect' algorithm, so that we
always examine the entire solution space until we find the minimum
solution. The algorithm can be tweaked via the \fB-d\fR, \fB-r\fR,
\fB-L\fR and \fB-l\fR options.

By default, the program will show the solution in a somewhat cryptic
format; for getting the actual Ganeti command list, use the \fB-C\fR
option.

\fBNote:\fR this program is somewhat deprecated; \fBhbal(1)\fR gives
usually much faster results, and a better cluster. It is recommended
to use this program only when \fBhbal\fR doesn't give a N+1 compliant
cluster.

.SS ALGORITHM

The algorithm works in multiple rounds, of increasing \fIdepth\fR,
until we have a solution.

First, before starting the solution computation, we compute all the
N+1-fail nodes and the instances they hold. These instances are
candidate for replacement (and only these!).

The program start then with \fIdepth\fR one (unless overridden via the
\fB-d\fR option), and at each round:
.RS 4
.TP 3
\(em
it tries to remove from the cluster as many instances as the current
depth in order to make the cluster N+1 compliant

.TP
\(em
then, for each of the possible instance combinations that allow this
(unless the total size is reduced via the \fB-r\fR option), it tries
to put them back on the cluster while maintaining N+1 compliance
.RE

It might be that at a given round, the results are:
.RS 4
.TP 3
\(em
no instance combination that can be put back; this means it is not
possible to make the cluster N+1 compliant with this number of
instances being moved, so we increase the depth and go on to the next
round
.TP
\(em
one or more successful result, in which case we take the one that has
as few changes as possible (by change meaning a replace-disks needed)
.RE

The main problem with the algorithm is that, being an exhaustive
search, the CPU time required grows very very quickly based on
depth. On a 20-node, 80-instances cluster, depths up to 5-6 are
quickly computed, and depth 10 could already take days.

The main factors that influence the run time are:
.RS 4
.TP 3
\(em
the removal depth; for each increase with one of the depth, we grow
the solution space by the number of nodes squared (since a new
instance can live any two nodes as primary/secondary, therefore
(almost) N times N); i.e., depth=1 will create a N^2 solution space,
depth two will make this N^4, depth three will be N^6, etc.

.TP
\(em
the removal depth again; for each increase in the depth, there will be
more valid removal sets, and the space of solutions increases linearly
with the number of removal sets
.RE

Therefore, the smaller the depth the faster the algorithm will be; it doesn't
seem like this algorithm will work for clusters of 100 nodes and many many
small instances (e.g. 256MB instances on 16GB nodes).

As an optimisation, since the algorithm is designed to prune the
search space as quickly as possible, is by luck we find a good
solution early at a given depth, then the other solutions which would
result in a bigger delta (the number of changes) will not be
investigated, and the program will finish fast. Since this is random
and depends on where in the full solution space the good solution will
be, there are two options for cutting down the time needed:
.RS 4
.TP 3
\(em
\fB-l\fR makes any solution that has delta lower than its parameter
succeed instantly; the default value for this parameter is zero, so
once we find a "perfect" solution we finish early

.TP
\(em
\fB-L\fR makes any solution with delta higher than its parameter being
rejected instantly (and not descend on the search tree); this can
reduce the depth of the search tree, with sometimes significant
speedups; by default, this optimization is not used
.RE

The algorithm also has some other internal optimisations:
.RS 4
.TP 3
\(em
when choosing where to place an instance in phase two, there are
N*(N-1) possible primary/secondary options; however, if instead of
iterating over all p * s pairs, we first determine the set of primary
nodes that can hold this instance (without failing N+1), we can cut
(N-1) secondary placements for each primary node removed; and since
this applies at every iteration of phase 2 it linearly decreases the
solution space, and on full clusters, this can mean a four-five times
reductions of solution space

.TP
\(em
since the number of solutions is very high even for smaller depths (on
the test data, depth=4 results in 1.8M solutions) we can't compare
them at the end, so at each iteration in phase 2 we only promote the
best solution out of our own set of solutions

.TP
\(em
since the placement of instances can only increase the delta of the
solution (placing a new instance will add zero or more replace-disks
steps), it means the delta will only increase while recursing during
phase 2; therefore, if we know at one point that we have a current
delta that is equal or higher to the delta of the best solution so
far, we can abort the recursion; this cuts a tremendous number of
branches; further promotion of the best solution from one removal set
to another can cut entire removal sets after a few recursions

.RE

.SH OPTIONS
The options that can be passed to the program are as follows:
.TP
.B -C, --print-commands
Print the command list at the end of the run. Without this, the
program will only show a shorter, but cryptic output.
.TP
.B -p, --print-nodes
Prints the before and after node status, in a format designed to allow
the user to understand the node's most important parameters.

The node list will contain these informations:
.RS
.TP
.B F
a character denoting the status of the node, with '-' meaning an
offline node, '*' meaning N+1 failure and blank meaning a good node
.TP
.B Name
the node name
.TP
.B t_mem
the total node memory
.TP
.B n_mem
the memory used by the node itself
.TP
.B i_mem
the memory used by instances
.TP
.B x_mem
amount memory which seems to be in use but cannot be determined why or
by which instance; usually this means that the hypervisor has some
overhead or that there are other reporting errors
.TP
.B f_mem
the free node memory
.TP
.B r_mem
the reserved node memory, which is the amount of free memory needed
for N+1 compliance
.TP
.B t_dsk
total disk
.TP
.B f_dsk
free disk
.TP
.B pcpu
the number of physical cpus on the node
.TP
.B vcpu
the number of virtual cpus allocated to primary instances
.TP
.B pri
number of primary instances
.TP
.B sec
number of secondary instances
.TP
.B p_fmem
percent of free memory
.TP
.B p_fdsk
percent of free disk
.TP
.B r_cpu
ratio of virtual to physical cpus
.RE

.TP
.BI "-n" nodefile ", --nodes=" nodefile
The name of the file holding node information (if not collecting via
RAPI), instead of the default \fInodes\fR file (but see below how to
customize the default value via the environment).

.TP
.BI "-i" instancefile ", --instances=" instancefile
The name of the file holding instance information (if not collecting
via RAPI), instead of the default \fIinstances\fR file (but see below
how to customize the default value via the environment).

.TP
.BI "-m" cluster
Collect data not from files but directly from the
.I cluster
given as an argument via RAPI. If the argument doesn't contain a colon
(:), then it is converted into a fully-built URL via prepending
https:// and appending the default RAPI port, otherwise it's
considered a fully-specified URL and is used unchanged.

.TP
.BI "-d" DEPTH ", --depth=" DEPTH
Start the algorithm directly at depth \fID\fR, so that we don't
examine lower depth. This will be faster if we know a solution is not
found a lower depths, and thus it's unneeded to search them.

.TP
.BI "-l" MIN-DELTA ", --min-delta=" MIN-DELTA
If we find a solution with delta lower or equal to \fIMIN-DELTA\fR,
consider this a success and don't examine further.

.TP
.BI "-L" MAX-DELTA ", --max-delta=" MAX-DELTA
If while computing a solution, it's intermediate delta is already
higher or equal to \fIMAX-DELTA\fR, consider this a failure and abort
(as if N+1 checks have failed).

.TP
.B -V, --version
Just show the program version and exit.

.SH EXIT STATUS

The exist status of the command will be zero, unless for some reason
the algorithm fatally failed (e.g. wrong node or instance data).

.SH ENVIRONMENT

If the variables \fBHTOOLS_NODES\fR and \fBHTOOLS_INSTANCES\fR are
present in the environment, they will override the default names for
the nodes and instances files. These will have of course no effect
when RAPI is used.

.SH BUGS

The program does not check its input data for consistency, and aborts
with cryptic errors messages in this case.

The algorithm doesn't deal with non-\fBdrbd\fR instances, and chokes
on input data which has such instances.

The algorithm doesn't know when it won't be possible to reach N+1
compliance at all, and will happily churn CPU for ages without
realising it won't reach a solution.

The algorithm is too slow.

The output format is not easily scriptable, and the program should
feed moves directly into Ganeti (either via RAPI or via a gnt-debug
input file).

.SH SEE ALSO
.BR hbal "(1), " hscan "(1), " ganeti "(7), " gnt-instance "(8), "
.BR gnt-node "(8)"

.SH "COPYRIGHT"
.PP
Copyright (C) 2009 Google Inc. Permission is granted to copy,
distribute and/or modify under the terms of the GNU General Public
License as published by the Free Software Foundation; either version 2
of the License, or (at your option) any later version.
.PP
On Debian systems, the complete text of the GNU General Public License
can be found in /usr/share/common-licenses/GPL.


NEWS
Version 2.0.1
  - added -H/-B startup parameters to gnt-instance, which will allow
    re-adding the start in single-user option (regression from 1.2)
  - the watcher writes the instance status to a file, to allow
    monitoring to report the instance status (from the master) based on
    cached results of the watcher's queries; while this can get stale if
    the watcher is being locked due to other work on the cluster, this
    is still an improvement
  - the watcher now also restarts the node daemon and the rapi daemon if
    they died
  - fixed the watcher to handle full and drained queue cases
  - hooks export more instance data in the environment, which helps if
    hook scripts need to take action based on the instance's properties
    (no longer need to query back into ganeti)
  - instance failovers when the instance is stopped do not check for
    free RAM, so that failing over a stopped instance is possible in low
    memory situations
  - rapi uses queries for tags instead of jobs (for less job traffic),
    and for cluster tags it won't talk to masterd at all but read them
    from ssconf
  - a couple of error handling fixes in RAPI
  - drbd handling: improved the error handling of inconsistent disks
    after resync to reduce the frequency of "there are some degraded
    disks for this instance" message
  - fixed a bug in live migration when DRBD doesn't want to reconnect
    (the error handling path called a wrong function name)

Version 2.0.0 final
  - no changes from rc5

Version 2.0 release candidate 5
  - fix a couple of bugs (validation, argument checks)
  - fix gnt-cluster getmaster on non-master nodes (regression)
  - some small improvements to RAPI and IAllocator
  - make watcher automatically start the master daemon if down

Version 2.0 release candidate 4
  - change the OS list to not require locks; this helps with big
    clusters
  - fix gnt-cluster verify and gnt-cluster verify-disks when the volume
    group is broken
  - “gnt-instance info”, without any arguments, doesn't run for all
    instances anymore; either pass ‘--all’ or pass the desired
    instances; this helps against mistakes on big clusters where listing
    the information for all instances takes a long time
  - miscellaneous doc and man pages fixes

Version 2.0 release candidate 3
  - Change the internal locking model of some gnt-node commands, in
    order to reduce contention (and blocking of master daemon) when
    batching many creation/reinstall jobs
  - Fixes to Xen soft reboot
  - No longer build documentation at 'make' time, instead distribute it
    in the archive, in order to reduce the need for the whole
    docbook/rst toolchains

Version 2.0 release candidate 2
  - Now the cfgupgrade scripts works and can upgrade 1.2.7 clusters to
    2.0
  - Fix watcher startup sequence, improves the behaviour of busy
    clusters
  - Some other fixes in cluster verify, replace-disks, gnt-instance add,
    gnt-cluster queue, KVM VNC bind address and other places
  - Some documentation fixes and updates

Version 2.0 release candidate 1
  - More documentation updates, now all docs should be more-or-less
    up-to-date
  - A couple of small fixes (mixed hypervisor clusters, offline nodes,
    etc.)
  - Added a customizable HV_KERNEL_ARGS hypervisor parameter (for Xen
    PVM and KVM)
  - Fix an issue related to $libdir/run/ganeti and cluster creation

Version 2.0 beta 2
  - Xen PVM and KVM have switched the default value for the instance
    root disk to the first partition on the first drive, instead of the
    whole drive; this means that the OS installation scripts must be
    changed accordingly
  - Man pages have been updated
  - RAPI has been switched by default to HTTPS, and the exported
    functions should all work correctly
  - RAPI v1 has been removed
  - Many improvements to the KVM hypervisor
  - Block device errors are now better reported
  - Many other bugfixes and small improvements

Version 2.0 beta 1
  - Version 2 is a general rewrite of the code and therefore the
    differences are too many to list, see the design document for 2.0 in
    the doc/ subdirectory for more details
  - In this beta version there is not yet a migration path from 1.2
    (there will be one in the final 2.0 release)
  - A few significant changes are:
    - all commands are executed by a daemon (ganeti-masterd) and the
      various gnt-* commands are just front-ends to it
    - all the commands are entered into, and executed from a job queue,
      see gnt-job(8)
    - the RAPI daemon supports read-write operations, secured by basic
      HTTP authentication on top of HTTPS
    - DRBD version 0.7 support has been removed, DRBD 8 is the only
      supported version (when migrating from Ganeti 1.2 to 2.0, you need
      to migrate to DRBD 8 first while still running Ganeti 1.2)
    - DRBD devices are using statically allocated minor numbers, which
      will be assigned to existing instances during the migration
      process
    - there is support for both Xen PVM and Xen HVM instances running on
      the same cluster
    - KVM virtualization is supported too
    - file-based storage has been implemented, which means that it is
      possible to run the cluster without LVM and DRBD storage, for
      example using a shared filesystem exported from shared storage
      (and still have live migration)

Version 1.2.7
  - Change the default reboot type in gnt-instance reboot to "hard"
  - Reuse the old instance mac address by default on instance import,
    if the instance name is the same.
  - Handle situations in which the node info rpc returns incomplete results
    (issue 46)
  - Add checks for tcp/udp ports collisions in gnt-cluster verify
  - Improved version of batcher:
    - state file support
    - instance mac address support
    - support for HVM clusters/instances
  - Add an option to show the number of cpu sockets and nodes in gnt-node list

hn1.hs
{-| Solver for N+1 cluster errors

-}

{-

Copyright (C) 2009 Google Inc.

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful, but
WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
02110-1301, USA.

-}

module Main (main) where

import Data.List
import Data.Function
import Monad
import System
import System.IO
import System.Console.GetOpt
import qualified System

import Text.Printf (printf)

import qualified Ganeti.HTools.Container as Container
import qualified Ganeti.HTools.Instance as Instance
import qualified Ganeti.HTools.Node as Node
import qualified Ganeti.HTools.Cluster as Cluster
import qualified Ganeti.HTools.CLI as CLI

-- | Command line options structure.
data Options = Options
    { optShowNodes   :: Bool
    , optShowCmds    :: Bool
    , optNodef       :: FilePath
    , optNodeSet     :: Bool     -- ^ The nodes have been set by options
    , optInstf       :: FilePath -- ^ Path to the instances file
    , optInstSet     :: Bool     -- ^ The insts have been set by options
    , optMinDepth    :: Int
    , optMaxRemovals :: Int
    , optMinDelta    :: Int
    , optMaxDelta    :: Int
    , optMaster      :: String
    , optShowVer     :: Bool     -- ^ Just show the program version
    , optShowHelp    :: Bool     -- ^ Just show the help
    } deriving Show

instance CLI.CLIOptions Options where
    showVersion = optShowVer
    showHelp    = optShowHelp

instance CLI.EToolOptions Options where
    nodeFile   = optNodef
    nodeSet    = optNodeSet
    instFile   = optInstf
    instSet    = optInstSet
    masterName = optMaster
    silent _   = False

-- | Default values for the command line options.
defaultOptions :: Options
defaultOptions    = Options
 { optShowNodes   = False
 , optShowCmds    = False
 , optNodef       = "nodes"
 , optNodeSet     = False
 , optInstf       = "instances"
 , optInstSet     = False
 , optMinDepth    = 1
 , optMaxRemovals = -1
 , optMinDelta    = 0
 , optMaxDelta    = -1
 , optMaster      = ""
 , optShowVer     = False
 , optShowHelp    = False
 }

{- | Start computing the solution at the given depth and recurse until
we find a valid solution or we exceed the maximum depth.

-}
iterateDepth :: Node.List
             -> [Instance.Instance]
             -> Int
             -> Int
             -> Int
             -> Int
             -> IO (Maybe Cluster.Solution)
iterateDepth nl bad_instances depth max_removals min_delta max_delta =
    let
        max_depth = length bad_instances
        sol = Cluster.computeSolution nl bad_instances depth
              max_removals min_delta max_delta
    in
      do
        printf "%d " depth
        hFlush stdout
        case sol `seq` sol of
          Nothing ->
              if depth > max_depth then
                  return Nothing
              else
                  iterateDepth nl bad_instances (depth + 1)
                               max_removals min_delta max_delta
          _ -> return sol

-- | Options list and functions
options :: [OptDescr (Options -> Options)]
options =
    [ Option ['p']     ["print-nodes"]
      (NoArg (\ opts -> opts { optShowNodes = True }))
      "print the final node list"
    , Option ['C']     ["print-commands"]
      (NoArg (\ opts -> opts { optShowCmds = True }))
      "print the ganeti command list for reaching the solution"
    , Option ['n']     ["nodes"]
      (ReqArg (\ f opts -> opts { optNodef = f, optNodeSet = True }) "FILE")
      "the node list FILE"
    , Option ['i']     ["instances"]
      (ReqArg (\ f opts -> opts { optInstf =  f, optInstSet = True }) "FILE")
      "the instance list FILE"
    , Option ['d']     ["depth"]
      (ReqArg (\ i opts -> opts { optMinDepth =  (read i)::Int }) "D")
      "start computing the solution at depth D"
    , Option ['r']     ["max-removals"]
      (ReqArg (\ i opts -> opts { optMaxRemovals =  (read i)::Int }) "R")
      "do not process more than R removal sets (useful for high depths)"
    , Option ['L']     ["max-delta"]
      (ReqArg (\ i opts -> opts { optMaxDelta =  (read i)::Int }) "L")
      "refuse solutions with delta higher than L"
    , Option ['l']     ["min-delta"]
      (ReqArg (\ i opts -> opts { optMinDelta =  (read i)::Int }) "L")
      "return once a solution with delta L or lower has been found"
    , Option ['m']     ["master"]
      (ReqArg (\ m opts -> opts { optMaster = m }) "ADDRESS")
      "collect data via RAPI at the given ADDRESS"
    , Option ['V']     ["version"]
      (NoArg (\ opts -> opts { optShowVer = True}))
      "show the version of the program"
    , Option ['h']     ["help"]
      (NoArg (\ opts -> opts { optShowHelp = True}))
      "show help"
    ]

-- | Main function.
main :: IO ()
main = do
  cmd_args <- System.getArgs
  (opts, args) <- CLI.parseOpts cmd_args "hn1" options
                  defaultOptions

  unless (null args) $ do
         hPutStrLn stderr "Error: this program doesn't take any arguments."
         exitWith $ ExitFailure 1

  (nl, il, csf) <- CLI.loadExternalData opts

  printf "Loaded %d nodes, %d instances\n"
             (Container.size nl)
             (Container.size il)

  when (length csf > 0) $ do
         printf "Note: Stripping common suffix of '%s' from names\n" csf

  let (bad_nodes, bad_instances) = Cluster.computeBadItems nl il
      min_depth = optMinDepth opts

  printf "Initial check done: %d bad nodes, %d bad instances.\n"
             (length bad_nodes) (length bad_instances)

  when (null bad_instances) $ do
         putStrLn "Happy time! Cluster is fine, no need to burn CPU."
         exitWith ExitSuccess

  when (length bad_instances < min_depth) $ do
         printf "Error: depth %d is higher than the number of bad instances.\n"
                min_depth
         exitWith $ ExitFailure 2

  let ini_cv = Cluster.compCV nl
  printf "Initial coefficients: overall %.8f, %s\n"
         ini_cv (Cluster.printStats nl)

  putStr "Computing solution: depth "
  hFlush stdout

  result <- iterateDepth nl bad_instances min_depth (optMaxRemovals opts)
            (optMinDelta opts) (optMaxDelta opts)
  let (min_d, solution) =
          case result of
            Just (Cluster.Solution a b) -> (a, reverse b)
            Nothing -> (-1, [])
  when (min_d == -1) $ do
         putStrLn "failed. Try to run with higher depth."
         exitWith $ ExitFailure 1

  printf "found.\n"

  let
      ns = Cluster.applySolution nl il solution
      fin_cv = Cluster.compCV ns

  printf "Final coefficients:   overall %.8f, %s\n"
         fin_cv
         (Cluster.printStats ns)

  printf "Solution (delta=%d):\n" $! min_d
  let (sol_strs, cmd_strs) = Cluster.printSolution ns il solution
  putStr $ unlines $ sol_strs
  when (optShowCmds opts) $
       do
         putStrLn ""
         putStrLn "Commands to run to reach the above solution:"
         putStr . Cluster.formatCmds . reverse $ cmd_strs

  when (optShowNodes opts) $
       do
         let (orig_mem, orig_disk) = Cluster.totalResources nl
             (final_mem, final_disk) = Cluster.totalResources ns
         putStrLn ""
         putStrLn "Final cluster status:"
         putStrLn $ Cluster.printNodes ns
         printf "Original: mem=%d disk=%d\n" orig_mem orig_disk
         printf "Final:    mem=%d disk=%d\n" final_mem final_disk
